apisix/plugins/api-breaker.lua

```diff
@@ -0,0 +1,221 @@
+-- (the "License"); you may not use this file except in compliance with
+-- the License.  You may obtain a copy of the License at
+--
+--     http://www.apache.org/licenses/LICENSE-2.0
+--
+-- Unless required by applicable law or agreed to in writing, software
+-- distributed under the License is distributed on an "AS IS" BASIS,
+-- WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+-- See the License for the specific language governing permissions and
+-- limitations under the License.
+--
+local plugin_name = "api-breaker"
+local ngx = ngx
+local math = math
+local ipairs = ipairs
+local error = error
+local core = require("apisix.core")
+
+local DEFAULT_EXPTIME = 300 -- TODO: user can config
+
+local shared_buffer = ngx.shared['plugin-'.. plugin_name]
+if not shared_buffer then
+    error("failed to get ngx.shared dict when load plugin " .. plugin_name)
+end
+
+
+local schema = {
+    type = "object",
+    properties = {
+        unhealthy_response_code = {
+            type = "integer",
+            minimum = 200,
+            maximum = 599,
+        },
+        unhealthy = {
+            type = "object",
+            properties = {
+                http_statuses = {
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 500,
+                        maximum = 599,
+                    },
+                    uniqueItems = true,
+                    default = {500}
+                },
+                failures = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        },
+        healthy = {
+            type = "object",
+            properties = {
+                http_statuses = {
+                    type = "array",
+                    minItems = 1,
+                    items = {
+                        type = "integer",
+                        minimum = 200,
+                        maximum = 499,
+                    },
+                    uniqueItems = true,
+                    default = {200, 206}
+                },
+                successes = {
+                    type = "integer",
+                    minimum = 1,
+                    default = 1,
+                }
+            }
+        }
+    },
+    required = {"unhealthy_response_code", "unhealthy", "healthy"},
+}
+
+
+local function is_unhealthy(unhealthy_status, upstream_status)
+    for _, unhealthy in ipairs(unhealthy_status) do
+        if unhealthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
+
+local function is_healthy(healthy_status, upstream_status)
+    for _, healthy in ipairs(healthy_status) do
+        if healthy == upstream_status then
+            return true
+        end
+    end
+
+    return false
+end
+
```

Comment:
Can be changed to a function, element in array

Reply:
For the convenience of reading